### PR TITLE
Bind NodeTypeConfig in all druid services

### PR DESCRIPTION
**Pull request by jon-wei**
Binds a NodeTypeConfig instance for all druid services. 

The historical, peon, and realtime services already had the binding, this PR adds a similar binding for the other services.

### CODE REVIEW

**Comment by nishantmonu51** [comment]
:+1:, If possible, would be nice to extend MainTest to verify it.

File: services/src/main/java/io/druid/cli/CliMiddleManager.java [hunk 1]
```diff
@@ -81,6 +82,8 @@ public void configure(Binder binder)
 
             IndexingServiceModuleHelper.configureTaskRunnerConfigs(binder);
 
+            binder.bind(NodeTypeConfig.class).toInstance(new NodeTypeConfig("middlemanager"));
```
Comment (nishantmonu51) [inline services/src/main/java/io/druid/cli/CliMiddleManager.java]: should this be "middleManager", consistent with the command name ?

File: services/src/main/java/io/druid/cli/CliInternalHadoopIndexer.java [hunk 2]
```diff
@@ -84,6 +85,8 @@ public void configure(Binder binder)
           {
             binder.bindConstant().annotatedWith(Names.named("serviceName")).to("druid/internal-hadoop-indexer");
             binder.bindConstant().annotatedWith(Names.named("servicePort")).to(0);
+            binder.bind(NodeTypeConfig.class).toInstance(new NodeTypeConfig("internal-hadoop-indexer"));
```
Comment (nishantmonu51) [inline services/src/main/java/io/druid/cli/CliInternalHadoopIndexer.java]: rename to "hadoop-indexer" , consistent with the command name ?

**Comment by drcrallen** [comment]
I'm a bit worried this pushes towards service type rules as opposed to a "role" based injection paradigm. 

Can you explain more on why this binding is required?

**Comment by jon-wei** [comment]
@drcrallen this is for an emitter extension I'm working on, the binding would be used to attach a "nodeType" field to the metrics sent out by the emitter

File: services/src/main/java/io/druid/cli/CliMiddleManager.java (hunk 1 above)
Comment (jon-wei) [inline services/src/main/java/io/druid/cli/CliMiddleManager.java]: sure, renamed to "middleManager"

File: services/src/main/java/io/druid/cli/CliInternalHadoopIndexer.java (hunk 2 above)
Comment (jon-wei) [inline services/src/main/java/io/druid/cli/CliInternalHadoopIndexer.java]: sure, renamed to "hadoop-indexer"

**Comment by jon-wei** [comment]
@nishantmonu51 added a nodeType check to MainTest

**Comment by fjy** [comment]
👍

**Comment by drcrallen** [comment]
👎  (please hold off on merging a bit) I'd like to see more information on the intended use case.

Here are the things I'm seeing as a result of this PR:

* There is both serviceName and NodeTypeConfig. the NodeTypeConfig is currently only used in `StorageNodeModule` for reporting server metadata. You are expanding its capacity here and I'm not sure that is the best way to do things
* It is entirely possible for an extension to inject a string value from a property. If this is for reporting purposes AND `serviceName` is insufficient, can you explain why having a customizable property string is also insufficient?

And as mentioned earlier, there hasn't been a good delineation of services vs capacities/modules. Different CLIs have a collection of capacities which are grouped together into the service. CLIs are extensible and custom extensions can have various implementations, and even different combinations of properties greatly change the behavior of the service (we have both fast and slow brokers for example, that behave very differently even though they are both CliBroker).

**Comment by drcrallen** [comment]
For example, does the assumption about binding `NodeTypeConfig` in the CLI still make sense and is required when there are hierarchical brokers that vary in how their query runner is bound rather than the CLI itself, or when there is are a bunch of spark tasks which do intermediate merging and then merge the final results in a driver? Or if there is an extension with [custom CLI impls](https://github.com/metamx/druid/tree/taskTier/extensions-core/tasktier/src/main/java/io/druid/cli)? What should the binding look like there?

**Comment by jon-wei** [comment]
@drcrallen

Thanks for the feedback, to elaborate on the use case, the nodeType would be used to provide necessary context for events in a metrics collection/analysis UI (since the meaning of a metric event depends on the type of node it came from).

serviceName can be overridden by users with their own naming conventions, so it's not suitable for this purpose.

A customizable property string would work but would require the user to manage type configuration for all their nodes, which I feel is more error-prone than a fixed property.

I'll close this PR, since the default serviceName is accessible to the emitter extension and that suffices for my purposes without needing this additional NodeTypeConfig setup.

**Pull request closed without merging**